### PR TITLE
Template part: attribute feature parity with group block

### LINKED ISSUE #20997: Template Parts should allow setting HTML tag
Related: #20200

This is a setting that would make a lot of sense in the template part inspector controls settings. We should also see if template parts should be by default a variation on the `group` block, so that they inherit color controls, etc.

cc @youknowriad

### LINKED ISSUE #22034: Template part controls should be identical to a group block
Currently, the template part controls look like this:
image

I would expect them to be the same as the group block including the ability to go full width and change background colors. The full width aspect would be good in the near future for testing, but I expect will no longer be necessary if we add some sort of grid system to the template editor.

Here's what I would expect:
image

@epiqueras What do you think?

### PR DESCRIPTION
**Pull request by noahtallen**
## Description
Closes #20997 and resolves #22034. Adds all attributes which currently exist on the group block to the template part block. (Including light block wrapper and tagName support.) 

**Note: These are just block attributes at the block level. Attributes are not persisted to a template part globally.** This means that a change to these block attributes only applies to where the template part block is used, likely on a **template** level, rather than a template part level.

## How has this been tested?
Locally in edit site.

## Screenshots 
Screenshot shows a custom HTML tagName, custom background color, and custom text color in editor and then on the front end. The margins are a separate issue :D

### Front end:
Screen Shot 2020-09-02 at 3 59 47 PM

### Editor:
Screen Shot 2020-09-02 at 3 59 01 PM

## Checklist:
- [ ] My code is tested.
- [x] My code follows the WordPress code style. 
- [x] My code follows the accessibility standards. 
- [x] My code has proper inline documentation. 
- [ ] I've included developer documentation if appropriate. 
- [x] I've updated all React Native files affected by any refactorings/renamings in this PR.

### CODE REVIEW

File: packages/block-library/src/template-part/index.php [hunk 1]
```diff
@@ -63,7 +63,8 @@ function render_block_core_template_part( $attributes ) {
 	}
 	$content = do_shortcode( $content );
 
-	return '<div>' . str_replace( ']]>', ']]&gt;', $content ) . '</div>';
+	$html_tag = $attributes['tagName'];
```
Comment (noahtallen) [inline packages/block-library/src/template-part/index.php:66]: is there some kind of escaping function we need to run here?

File: packages/block-library/src/template-part/edit/inner-blocks.js [hunk 2]
```diff
@@ -19,6 +19,7 @@ export default function TemplatePartInnerBlocks( {
 			value={ blocks }
 			onInput={ onInput }
 			onChange={ onChange }
+			__experimentalTagName="div"
```
Comment (Addison-Stavlo) [inline packages/block-library/src/template-part/edit/inner-blocks.js:22]: Why was this necessary for the inner-blocks?

File: packages/block-library/src/template-part/edit/index.js [hunk 3]
```diff
@@ -59,8 +62,9 @@ export default function TemplatePartEdit( {
 
 	if ( postId ) {
 		// Part of a template file, post ID already resolved.
+		const BlockWrapper = Block[ tagName ];
 		return (
-			<>
+			<BlockWrapper>
```
Comment (Addison-Stavlo) [inline packages/block-library/src/template-part/edit/index.js:67]: We may need to add a wrapper like this to the placeholder as well.

File: packages/block-library/src/template-part/index.php (hunk 1 above)
Comment (Addison-Stavlo) [reply packages/block-library/src/template-part/index.php:66]: I think `esc_attr` may still be useful in this case?

**Review by Addison-Stavlo — COMMENTED** [review]
Tested this out and it works well minus a couple things:

1. The placeholder component is now taking up full width in the editor.  This is probably a result of switching to the light wrapper without adding a wrapper around the Placeholders return as well.

Screen Shot 2020-09-02 at 8 22 18 PM

2.  I can't find a way to change the tagName?  I am assuming this is something we will need to add manually.

**Comment by noahtallen** [comment]
> I can't find a way to change the tagName

I don't think this is supported by any blocks right now. I think you just define it in the block HTML

File: packages/block-library/src/template-part/edit/index.js (hunk 3 above)
Comment (noahtallen) [reply packages/block-library/src/template-part/edit/index.js:67]: Yeah, we could do that. I didn't think it mattered in the place holder state :)

**Comment by Addison-Stavlo** [comment]
> I don't think this is supported by any blocks right now. I think you just define it in the block HTML

Ah interesting, I think we should definitely add a way to set this in a followup... it seems like it should be necessary.  🤔

File: packages/block-library/src/template-part/edit/inner-blocks.js (hunk 2 above)
Comment (noahtallen) [reply packages/block-library/src/template-part/edit/inner-blocks.js:22]: It tells the inner blocks component to return the inner blocks directly as an array instead of putting them inside its own wrapper. So we get this:

```
<header>
  <innerBlock1 />
  <innerBlock2 />
</header>
```

and not this:
```
<header>
  <innerBlocksDiv>
    <innerBlock1 />
    <innerBlock2 />
  </innerBlocksDiv>
</header>
```

It's weird because it's just a boolean check and "div" doesn't actually do anything. But this seems to be the format used by the group block and others.

**Comment by noahtallen** [comment]
Yeah, true. I wonder if there are existing issues for that for the group block and others

File: packages/block-library/src/template-part/edit/inner-blocks.js (hunk 2 above)
Comment (Addison-Stavlo) [reply packages/block-library/src/template-part/edit/inner-blocks.js:22]: > It's weird because it's just a boolean check and "div" doesn't actually do anything. But this seems to be the format used by the group block and others.

Hah, that is weird...

File: packages/block-library/src/template-part/index.php (hunk 1 above)
Comment (noahtallen) [reply packages/block-library/src/template-part/index.php:66]: Yeah. I guess, if it _does_ end up escaping anything, it will be an invalid HTML tag, but there would be bigger issues in that case anyways :P

**Comment by noahtallen** [comment]
> The placeholder component is now taking up full width in the editor. This is probably a result of switching to the light wrapper without adding a wrapper around the Placeholders return as well.

Added the wrapper there and it seems to be fixed ✅

**Review by Addison-Stavlo — APPROVED** [review]
> Added the wrapper there and it seems to be fixed ✅

Works for me. 😁

**Comment by youknowriad** [comment]
According to the screenshot, it seems we're missing the `has-background` style (padding) for this block?

**Comment by noahtallen** [comment]
I think I do see it on the front end (from that screenshot):

Screen Shot 2020-09-03 at 10 26 28 AM

Are there other classnames that might add margin there? Maybe the group block has some default margin/padding styles we need to add to the template part block?

**Comment by noahtallen** [comment]
@youknowriad I did some investigation, and those styles are defined here in the group block, so they won't apply to template parts out of the box:

https://github.com/WordPress/gutenberg/blob/0b93ad47797a1ba2cea3c83a7f973d33abcd351f/packages/block-library/src/group/theme.scss#L1-L10

I'm not sure how to solve this, since it seems there are several styles scattered around to support padding/margin for the group block. I'm not sure exactly what we should change. For example, with the theme I'm testing, it seems the group block padding is ultimately set by the theme, and the theme does not set anything for template part blocks yet.

**Comment by noahtallen** [comment]
> I'm not sure how to solve this

I've copied over the theme.scss file from the group block. So this should provide good baseline styles. but it still seems like padding/margin is primarily something themes need to code?

At any rate, I'll merge what we have since it seems to work well. As always, I'm happy to address feedback in followups
